File: ChromeHistoryReader/Form1.cs
using System;
using System.Drawing;
using System.Windows.Forms;

namespace ChromeHistoryReader
{
	public partial class Form1 : Form
	{
		private HistoryDataProvider _dataProvider;
		private BindingSource _source;

		public Form1(HistoryDataProvider dataProvider)
		{
			this._dataProvider = dataProvider;
			_source = new BindingSource();
			InitializeComponent();
			this.dataGridView1.DataSource = _source;
		}

		private void button1_Click(object sender, EventArgs e)
		{
			RefreshContent();
		}

		private void dataGridView1_MouseDown(object sender, MouseEventArgs e)
		{
			if (e.Button == MouseButtons.Right)
			{
				var hti = dataGridView1.HitTest(e.X, e.Y);
				dataGridView1.ClearSelection();
				if (dataGridView1.Rows.Count > 0 && hti.RowIndex >= 0)
				{
					dataGridView1.Rows[hti.RowIndex].Selected = true;

					this.dataGridView1.ContextMenu.Show(dataGridView1, new Point(e.X, e.Y));
				}
			}
		}

		private void MenuItem_Click(object sender, EventArgs e)
		{
			var rowToDelete = this.dataGridView1.Rows.GetFirstRow(DataGridViewElementStates.Selected);
			var item = (HistoryRecord)dataGridView1.Rows[rowToDelete].DataBoundItem;

			if (item != null)
			{
				var rowWasDeleted = _dataProvider.DeleteHistoryItem(item.Id);
				if (rowWasDeleted)
				{
					_source.RemoveAt(rowToDelete);
				}
			}
			this.dataGridView1.ClearSelection();
		}

		private async void RefreshContent()
		{
			var records = await _dataProvider.GetHistoryRecords();
			_source.DataSource = records;
			this.dataGridView1.ClearSelection();
		}
	}
}

File: ChromeHistoryReaderVB/Form1.vb
Partial Class Form1

    Private button1 As Button
    Private tableLayoutPanel1 As TableLayoutPanel
    Private dataGridView1 As DataGridView

    Private _dataProvider As HistoryDataProvider
    Private _source As BindingSource

    Public Sub New(dataProvider As HistoryDataProvider)
        Me._dataProvider = dataProvider
        _source = New BindingSource()
        InitializeComponent()
        Me.dataGridView1.DataSource = _source
    End Sub

    Private Sub button1_Click(sender As Object, e As EventArgs)
        RefreshContent()
    End Sub

    Private Sub dataGridView1_MouseDown(sender As Object, e As MouseEventArgs)
        If e.Button = MouseButtons.Right Then
            Dim hti = dataGridView1.HitTest(e.X, e.Y)
            dataGridView1.ClearSelection()
            If dataGridView1.Rows.Count > 0 AndAlso hti.RowIndex >= 0 Then
                dataGridView1.Rows(hti.RowIndex).Selected = True

                Me.dataGridView1.ContextMenu.Show(dataGridView1, New Point(e.X, e.Y))
            End If
        End If
    End Sub

    Private Sub MenuItem_Click(sender As Object, e As EventArgs)
        Dim rowToDelete = Me.dataGridView1.Rows.GetFirstRow(DataGridViewElementStates.Selected)
        Dim item = DirectCast(dataGridView1.Rows(rowToDelete).DataBoundItem, HistoryRecord)

        If item IsNot Nothing Then
            Dim rowWasDeleted = _dataProvider.DeleteHistoryItem(item.Id)
            If rowWasDeleted Then
                _source.RemoveAt(rowToDelete)
            End If
        End If
        Me.dataGridView1.ClearSelection()
    End Sub

    Private Async Sub RefreshContent()
        Dim records = Await _dataProvider.GetHistoryRecords()
        _source.DataSource = records
        Me.dataGridView1.ClearSelection()
    End Sub
End Class
